multiplex.go

```diff
@@ -167,94 +164,61 @@ func (mp *Multiplex) sendMsg(ctx context.Context, header uint64, data []byte) er
 }
 
 func (mp *Multiplex) handleOutgoing() {
-	for {
-		select {
-		case <-mp.shutdown:
-			return
-
-		case data := <-mp.writeCh:
-			err := mp.writeMsg(data)
-			if err != nil {
-				// the connection is closed by this time
-				log.Warningf("error writing data: %s", err.Error())
-				return
-			}
-		}
+	defer mp.closeNoWait()
+	if err := mp.sendLoop(); err != nil {
+		log.Info("send loop exited with: %s", err)
 	}
 }
 
-func (mp *Multiplex) writeMsg(data []byte) error {
-	if len(data) >= 512 {
-		return mp.doWriteMsg(data)
-	}
+func (mp *Multiplex) sendLoop() error {
+	writeTimer := time.NewTimer(WriteCoalesceDelay)
 
-	buf := pool.Get(4096)
-	defer pool.Put(buf)
+	writer := pool.Writer{W: mp.con}
 
-	n := copy(buf, data)
-	pool.Put(data)
+	for {
+		var data []byte
 
-	if !mp.writeTimerFired {
-		if !mp.writeTimer.Stop() {
-			<-mp.writeTimer.C
+		// Early exit for shutdown.
+		select {
+		case <-mp.shutdown:
+			return nil
+		default:
 		}
-	}
-	mp.writeTimer.Reset(WriteCoalesceDelay)
-	mp.writeTimerFired = false
 
-	for {
+		// Write data if we have some immediately available
 		select {
 		case data = <-mp.writeCh:
-			wr := copy(buf[n:], data)
-			if wr < len(data) {
-				// we filled the buffer, send it
-				err := mp.doWriteMsg(buf)
-				if err != nil {
-					pool.Put(data)
+		default:
+			// Otherwise, wait until the end to the coalesce delay.
+			select {
+			case data = <-mp.writeCh:
+			case <-mp.shutdown:
+				return nil
+			case <-writeTimer.C:
```

Comment:
so wait, what happens if we got no data when the timer fires?
I think we are going to end up getting lots of interrupts from this for idle connections.

Reply:
We go to line 206 and wait.

Reply:
we'll get a fired timer then if the wait is arbitrary before we are doing any coalescing.

Reply:
Ah, I see. Yeah, we'll "flush" no data. But we should only get to this point if:

1. The connection is new (we'll have no data to flush so flush will be a noop).
2. We've read some data off the data chan.

Reply:
I am more concerned about waiting a bit, having the timer fire while we are waiting for data, and then see it as immediately fired without actually coalescing anything.